File: 240703/네 방향 탈출 가능 여부 판별하기/determine-escapableness-with-4-ways.cpp
#include <iostream>
#include <algorithm>
#include <queue>
#include <tuple>
using namespace std;
#define MAX_NM 100
#define DIR_NUM 4

int n, m;
int grid[MAX_NM][MAX_NM];
bool visited[MAX_NM][MAX_NM];
queue<pair<int, int>> q;

bool InRange(int x, int y) {
    return 0<=x && x<n && 0<=y && y<m;
}

bool CanGO(int x, int y) {
    if(!InRange(x, y)) return false;
    if(visited[x][y] || grid[x][y] == 0) return false;

    return true;
}

void BFS() {
    int dx[DIR_NUM] = {-1, 1, 0, 0};
    int dy[DIR_NUM] = {0, 0, -1, 1};

    while(!q.empty()) {
        int x, y;
        tie(x, y) = q.front();
        q.pop();

        for(int d=0; d<DIR_NUM; d++) {
            int nx = x+dx[d], ny = y+dy[d];
            if(CanGO(nx, ny)) {
                q.push(make_pair(nx, ny));
                visited[nx][ny] = 1;
            }
        }
    }
}

int main() {
    // 여기에 코드를 작성해주세요.
    cin >> n>>m;
    for(int i=0; i<n; i++) {
        for(int j=0; j<m; j++) {
            cin >> grid[i][j];
        }
    }

    q.push(make_pair(0, 0));
    visited[0][0] = 1;
    BFS();

    cout << visited[n-1][m-1];
    return 0;
}

File: 250130/네 방향 탈출 가능 여부 판별하기/determine-escapableness-with-4-ways.py
from collections import deque

n, m = tuple(map(int, input().split()))

grid = [
    list(map(int, input().split()))
    for _ in range(n)
]

visited = [[0 for _ in range(m)] for _ in range(n)]

dxs, dys = [-1, 1, 0, 0], [0, 0, -1, 1]

q = deque()

def inRange(x, y):
    return 0<=x<n and 0<=y<m

def can_go(x, y):
    return inRange(x, y) and grid[x][y] and not visited[x][y]

def bfs():
    while q:
        x, y = q.popleft()

        for dx, dy in zip(dxs, dys):
            nx, ny = x+dx, y+dy

            if can_go(nx, ny):
                q.append((nx, ny))
                visited[nx][ny] = 1

q.append((0, 0))
visited[0][0] = 1

bfs()

print(visited[n-1][m-1])